### PR TITLE
extract browser related properties

### PR DESCRIPTION
**Pull request by mimmi20**
With this PR browser related properties are moved to their own file and removed them from the useragents section.

As a side effect caused by the script I used to change all source files, the properties inside the sections now are sorted, the versions and the platforms are sorted too.

@asgrim, @jaydiablo Would you like to have this script as a command (in a seperate PR)? 

This PR requires PR #1746 to be merged before.

### CODE REVIEW

**Comment by mmorel-35** [comment]
The browser file is really huge. Would it be possible to split it a way that make maintenance easy?

File: tests/issues/issue-996.php [hunk 1]
```diff
@@ -111,7 +109,6 @@
             'JavaScript' => true,
             'VBScript' => false,
             'JavaApplets' => false,
-            'isSyndicationReader' => false,
```
Comment (jaydiablo) [inline tests/issues/issue-996.php]: Why is this property being removed from tests? Is it one of the deprecated properties we're removing as part of #1433?

Are the tests modified in any other way? There are a lot of files to expand here to see what was modified, especially if the only change is the removal of this property.

Comment (mimmi20) [reply tests/issues/issue-996.php]: >  Is it one of the deprecated properties we're removing as part of #1433?

Yes it is.

> Are the tests modified in any other way?

There are some test cases where I added a seperate tests for the lite mode. (These are the test files with added lines)

Comment (mimmi20) [reply tests/issues/issue-996.php]: >> Is it one of the deprecated properties we're removing as part of #1433?

Removing this property was a mistake. I looked not properly enough at the deprecated list.

File: tests/issues/issue-000-invalid-versions.php [hunk 2]
```diff
@@ -316,146 +802,490 @@
...
             'RenderingEngine_Name' => 'Gecko',
             'RenderingEngine_Version' => 'unknown',
+            'RenderingEngine_Maker' => 'Mozilla Foundation',
         ],
         'lite' => false,
         'standard' => false,
         'full' => true,
     ],
-    // Same as above
+    // Firefox 4.0 and up don't suppport OSX 10.4 (https://www.mozilla.org/en-US/firefox/4.0/system-requirements/)
```
Comment (jaydiablo) [inline tests/issues/issue-000-invalid-versions.php]: This comment got moved.

File: tests/issues/issue-000-invalid-versions.php [hunk 3]
```diff
@@ -8,12 +8,39 @@
         'properties' => [
             'Comment' => 'Firefox Generic',
             'Browser' => 'Firefox',
+            'Browser_Type' => 'Browser',
```
Comment (jaydiablo) [inline tests/issues/issue-000-invalid-versions.php]: I don't think we should add all of these properties to the tests in this file.

File: tests/issues/issue-000-invalids.php [hunk 4]
```diff
@@ -1,60 +1,395 @@
 <?php
```
Comment (jaydiablo) [inline tests/issues/issue-000-invalids.php]: Please revert the changes to this file.

File: tests/issues/issue-000-invalid-versions.php (hunk 3 above)
Comment (mimmi20) [reply tests/issues/issue-000-invalid-versions.php]: reverted

File: tests/issues/issue-000-invalid-versions.php (hunk 2 above)
Comment (mimmi20) [reply tests/issues/issue-000-invalid-versions.php]: reverted

File: tests/issues/issue-000-invalids.php (hunk 4 above)
Comment (mimmi20) [reply tests/issues/issue-000-invalids.php]: reverted

File: src/Browscap/Data/Expander.php [hunk 5]
```diff
@@ -71,6 +72,7 @@ private function resetPatternId() : void
             'useragent' => '',
             'platform' => '',
             'device' => '',
+            'browser' => '',
```
Comment (jaydiablo) [inline src/Browscap/Data/Expander.php]: I'm not sure this is necessary to have the browser (or the engine) in the pattern id.

They're not used in the `getPatternId` method, and not included in coverage reports.

If you really do think we should add them (because you want to expand coverage to process the browser and engine files), then you should make sure you add "engine" here too (since you're setting it below).

File: src/Browscap/Data/Expander.php [hunk 6]
```diff
@@ -160,6 +168,23 @@ private function parseUserAgent(UserAgent $uaData, bool $lite, bool $standard, i
             $deviceProperties = [];
         }
 
+        if (null !== $uaData->getBrowser()) {
+            $this->patternId['browser'] = $uaData->getBrowser();
```
Comment (jaydiablo) [inline src/Browscap/Data/Expander.php]: Again, I don't think this is necessary.

File: src/Browscap/Data/Expander.php [hunk 7]
```diff
@@ -237,8 +270,9 @@ private function parseChildren(string $ua, array $uaDataChild, bool $lite = true
                 $uaBase = str_replace('#PLATFORM#', $platformProperties->getMatch(), $uaDataChild['match']);
 
                 if (array_key_exists('engine', $uaDataChild)) {
-                    $engine           = $this->collection->getEngine($uaDataChild['engine']);
-                    $engineProperties = $engine->getProperties();
+                    $engine                    = $this->collection->getEngine($uaDataChild['engine']);
+                    $engineProperties          = $engine->getProperties();
+                    $this->patternId['engine'] = $uaDataChild['engine'];
```
Comment (jaydiablo) [inline src/Browscap/Data/Expander.php]: Not necessary (imo), but if you add here, need to make sure it's in the reset array.

File: src/Browscap/Data/Expander.php [hunk 8]
```diff
@@ -277,8 +327,9 @@ private function parseChildren(string $ua, array $uaDataChild, bool $lite = true
             $properties = ['Parent' => $ua, 'lite' => $lite, 'standard' => $standard];
 
             if (array_key_exists('engine', $uaDataChild)) {
-                $engine           = $this->collection->getEngine($uaDataChild['engine']);
-                $engineProperties = $engine->getProperties();
+                $engine                    = $this->collection->getEngine($uaDataChild['engine']);
+                $engineProperties          = $engine->getProperties();
+                $this->patternId['engine'] = $uaDataChild['engine'];
```
Comment (jaydiablo) [inline src/Browscap/Data/Expander.php]: Another case.

File: src/Browscap/Data/Factory/BrowserFactory.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+declare(strict_types = 1);
+namespace Browscap\Data\Factory;
+
+use Assert\Assertion;
+use Browscap\Data\Browser;
+
+class BrowserFactory
+{
+    /**
+     * Load a engines.json file and parse it into the platforms data array
```
Comment (jaydiablo) [inline src/Browscap/Data/Factory/BrowserFactory.php]: Not loading engines.json

File: src/Browscap/Data/Factory/UserAgentFactory.php [hunk 10]
```diff
@@ -10,16 +10,11 @@ class UserAgentFactory
...
      * @param bool    $isCore
      *
-     * @throws \RuntimeException If the file does not exist or has invalid JSON
-     * @throws \LogicException   If required attibutes are missing in the division
-     * @throws \LogicException
-     *
      * @return UserAgent[]
      */
-    public function build(array $userAgentsData, array $versions, bool $isCore) : array
+    public function build(array $userAgentsData, bool $isCore) : array
```
Comment (jaydiablo) [inline src/Browscap/Data/Factory/UserAgentFactory.php]: Since the signature is changing here, did you search for all cases where this would have been called?

Also, should be documented in the Readme?

File: src/Browscap/Data/Validator/DivisionDataValidator.php [hunk 11]
```diff
@@ -413,6 +427,26 @@ public function checkDeviceProperties(array $properties, string $message) : void
         }
     }
 
+    /**
+     * checks if device properties are set inside a properties array
```
Comment (jaydiablo) [inline src/Browscap/Data/Validator/DivisionDataValidator.php]: "checks if **browser** properties"

File: tests/BrowscapTest/Data/DataCollectionTest.php [hunk 12]
```diff
@@ -26,21 +26,6 @@ public function setUp() : void
         $this->object = new DataCollection($logger);
     }
 
-    private function getPlatformsJsonFixture()
```
Comment (jaydiablo) [inline tests/BrowscapTest/Data/DataCollectionTest.php]: I assume these are unused now (I think I recall you removing these fixture files a while back).

File: src/Browscap/Data/Expander.php (hunk 5 above)
Comment (mimmi20) [reply src/Browscap/Data/Expander.php]: fixed

File: src/Browscap/Data/Expander.php (hunk 6 above)
Comment (mimmi20) [reply src/Browscap/Data/Expander.php]: fixed

File: src/Browscap/Data/Expander.php (hunk 7 above)
Comment (mimmi20) [reply src/Browscap/Data/Expander.php]: fixed

File: src/Browscap/Data/Expander.php (hunk 8 above)
Comment (mimmi20) [reply src/Browscap/Data/Expander.php]: fixed

File: src/Browscap/Data/Factory/BrowserFactory.php (hunk 9 above)
Comment (mimmi20) [reply src/Browscap/Data/Factory/BrowserFactory.php]: fixed

File: src/Browscap/Data/Validator/DivisionDataValidator.php (hunk 11 above)
Comment (mimmi20) [reply src/Browscap/Data/Validator/DivisionDataValidator.php]: fixed

File: src/Browscap/Data/Factory/UserAgentFactory.php (hunk 10 above)
Comment (mimmi20) [reply src/Browscap/Data/Factory/UserAgentFactory.php]: added to Readme

File: tests/BrowscapTest/Data/DataCollectionTest.php (hunk 12 above)
Comment (mimmi20) [reply tests/BrowscapTest/Data/DataCollectionTest.php]: yes, they were unused.

**Comment by jaydiablo** [comment]
@mimmi20 regarding that script, I think it could be useful for sure.

I have a number of different scripts I use when doing updates, I could look at moving them into commands as well if they seem useful to others.
